rpaas/manager.go

```diff
@@ -0,0 +1,214 @@
+package rpaas
+
+import (
+	"bytes"
+	"context"
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/rsa"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"sync"
+
+	nginxv1alpha1 "github.com/tsuru/nginx-operator/pkg/apis/nginx/v1alpha1"
+	"github.com/tsuru/rpaas-operator/pkg/apis/extensions/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	k8sErrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type RpaasManager interface {
+	UpdateCertificate(string, *tls.Certificate) error
+}
+
+var rpaasManagerHolder = &struct {
+	sync.Mutex
+	m RpaasManager
+}{}
+
+func GetRpaasManager() RpaasManager {
+	rpaasManagerHolder.Lock()
+	defer rpaasManagerHolder.Unlock()
+	return rpaasManagerHolder.m
+}
+
+func SetRpaasManager(m RpaasManager) {
+	rpaasManagerHolder.Lock()
+	defer rpaasManagerHolder.Unlock()
+	rpaasManagerHolder.m = m
+}
+
+type k8sRpaasManager struct {
+	cli       client.Client
+	namespace string
+}
+
+func NewK8SRpaasManager(cli client.Client) RpaasManager {
+	return &k8sRpaasManager{
+		cli:       cli,
+		namespace: metav1.NamespaceDefault,
+	}
+}
+
+func (m *k8sRpaasManager) UpdateCertificate(instance string, c *tls.Certificate) error {
+	if c == nil {
+		return errors.New("manager: could not update certificate to nil")
+	}
+	rpaasInstance, err := m.getRpaasInstanceByName(instance)
+	if err != nil {
+		return err
+	}
+	secret, err := m.getCertificateSecret(rpaasInstance, v1alpha1.CertificateNameDefault)
+	if err != nil {
+		if k8sErrors.IsNotFound(err) {
+			secret, err = m.createCertificateSecret(rpaasInstance, v1alpha1.CertificateNameDefault, c)
+			if err != nil {
+				return err
+			}
+			certs := map[v1alpha1.CertificateName]nginxv1alpha1.TLSSecret{
+				v1alpha1.CertificateNameDefault: *newTLSSecret(secret, v1alpha1.CertificateNameDefault),
+			}
+			_, err = m.updateCertificates(rpaasInstance, certs)
```

Comment:
The struct returned by both `updateCertificates` and `updateCertificateSecret` is never actually used, shouldn't we return only an error in these cases?

Reply:
Less is more. Fixing at 3, 2...